extensions-contrib/scan-query/src/main/java/io/druid/query/scan/ScanQueryRunnerFactory.java

```diff
@@ -0,0 +1,126 @@
+/*
+ * Licensed to Metamarkets Group Inc. (Metamarkets) under one
+ * or more contributor license agreements. See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership. Metamarkets licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package io.druid.query.scan;
+
+import com.google.common.base.Function;
+import com.google.inject.Inject;
+import io.druid.common.utils.JodaUtils;
+import io.druid.java.util.common.ISE;
+import io.druid.java.util.common.guava.Sequence;
+import io.druid.java.util.common.guava.Sequences;
+import io.druid.query.Query;
+import io.druid.query.QueryContextKeys;
+import io.druid.query.QueryRunner;
+import io.druid.query.QueryRunnerFactory;
+import io.druid.query.QueryToolChest;
+import io.druid.segment.Segment;
+
+import java.util.Map;
+import java.util.concurrent.ExecutorService;
+
+public class ScanQueryRunnerFactory implements QueryRunnerFactory<ScanResultValue, ScanQuery>
+{
+  private final ScanQueryQueryToolChest toolChest;
+  private final ScanQueryEngine engine;
+
+  @Inject
+  public ScanQueryRunnerFactory(
+      ScanQueryQueryToolChest toolChest,
+      ScanQueryEngine engine
+  )
+  {
+    this.toolChest = toolChest;
+    this.engine = engine;
+  }
+
+  @Override
+  public QueryRunner<ScanResultValue> createRunner(Segment segment)
+  {
+    return new ScanQueryRunner(engine, segment);
+  }
+
+  @Override
+  public QueryRunner<ScanResultValue> mergeRunners(
+      ExecutorService queryExecutor,
+      final Iterable<QueryRunner<ScanResultValue>> queryRunners
+  )
+  {
+    // in single thread and in jetty thread instead of processing thread
```

Comment:
with processing thread, you would reserve a ByteBuffer but beyond that memory usage could stays as your current implementation.
ok, i see the argument about not taking up a processing thread for long time and letting them be available for other regular queries... then query cancellation does not work right now. but, it does not matter if you only do scan query directly on the historical.